kopf/structs/diffs.py

```diff
@@ -19,6 +19,16 @@ def resolve(d: Mapping, path: DiffPath):
     return result
 
 
+def reduce_iter(d: Diff, path: DiffPath) -> Generator[DiffItem, None, None]:
+    for op, field, old, new in d:
+        if field[:len(path)] == path:
+            yield (op, field[len(path):], old, new)
+
+
+def reduce(d: Diff, path: DiffPath) -> Diff:
+    return type(d)(reduce_iter(d, path))
```

Comment:
so Diff is callable?

Reply:
`Diff` is not actually a type, it is a declaration of a type with the new Python3 type hinting and annotations (see [here](https://docs.python.org/3/library/typing.html)) — i.e. an "interface" in OOP terms.

The real types matching `Diff` are `list`, `tuple`, and other [Sequences](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes). They are indeed callable (as any classes); and when called, they create a new instance from the provided argument(s).

Reply:
This pattern — `type(oldval)(newval)` — is basically about creating an instance of the same type as was fed into the function, without many if's.